From 0a9dc3ddad2d10865f1fd7d6159a5737851ff71c Mon Sep 17 00:00:00 2001
From: Pratik Pachange <ppachang@qti.qualcomm.com>
Date: Mon, 12 May 2025 18:09:06 +0530
Subject: [PATCH] v4l2object: Provide padding requirements during propose
 allocation

If the driver has size alignment requirement, suggest the
difference between aligned size required by diver and the actual
size of the frame as padding requirement to the upstream element.

This will ensure that the buffer size allocation is as per the
driver requirement.

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/8971>

Upstream-Status: Backport [https://gitlab.freedesktop.org/gstreamer/gstreamer/-/commit/8b12aea4c7bb50d1bb85e6c90f16489f262187db ]

Signed-off-by: Raja Ganapathi Busam <rbusam@qti.qualcomm.com>
---
 sys/v4l2/gstv4l2object.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index d823337..58634a9 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -6060,10 +6060,11 @@ gst_v4l2_object_propose_allocation (GstV4l2Object * obj, GstQuery * query)
 {
   GstBufferPool *pool = NULL;
   /* we need at least 2 buffers to operate */
-  guint size, min, max;
+  guint size, min, max, aligned_size = 0;
   GstCaps *caps;
   gboolean need_pool;
   GstStructure *allocation_meta = NULL;
+  GstAllocationParams allocation_params;

   /* Set defaults allocation parameters */
   size = GST_V4L2_SIZE (obj);
@@ -6113,6 +6114,24 @@ gst_v4l2_object_propose_allocation (GstV4l2Object * obj, GstQuery * query)

   min = MAX (obj->min_buffers, GST_V4L2_MIN_BUFFERS (obj));

+  /* If the driver has any size alignment requirements, suggest the difference
+   * between aligned size and actual size as extra padding to the upstream
+   * element */
+  if (V4L2_TYPE_IS_MULTIPLANAR (obj->type)) {
+    /* FIXME */
+  } else {
+    aligned_size = obj->format.fmt.pix.sizeimage;
+  }
+
+  gst_allocation_params_init (&allocation_params);
+  allocation_params.padding = aligned_size - size;
+
+  if (allocation_params.padding) {
+    gst_query_add_allocation_param (query, NULL, &allocation_params);
+    /* Update size to aligned size required by driver */
+    size = aligned_size;
+  }
+
   gst_query_add_allocation_pool (query, pool, size, min, max);

   if (obj->align.padding_top || obj->align.padding_bottom ||
--
2.34.1
